homeassistant/components/bluetooth/passive_update_coordinator.py

```diff
@@ -0,0 +1,304 @@
+            for entity_key, description in data.entity_descriptions.items():
+                if entity_key not in created:
+                    entities.append(entity_class(self, entity_key, description))
+                    created.add(entity_key)
+            if entities:
+                async_add_entites(entities)
+
+        return self.async_add_listener(_async_add_or_update_entities)
+
+    @callback
+    def async_add_listener(
+        self,
+        update_callback: Callable[[PassiveBluetoothDataUpdate[_T] | None], None],
+    ) -> Callable[[], None]:
+        """Listen for all updates."""
+
+        @callback
+        def remove_listener() -> None:
+            """Remove update listener."""
+            self._listeners.remove(update_callback)
+
+        self._listeners.append(update_callback)
+        return remove_listener
+
+    @callback
+    def async_add_entity_key_listener(
+        self,
+        update_callback: Callable[[PassiveBluetoothDataUpdate[_T] | None], None],
+        entity_key: PassiveBluetoothEntityKey | None = None,
+    ) -> Callable[[], None]:
+        """Listen for updates by device key."""
+
+        @callback
+        def remove_listener() -> None:
+            """Remove update listener."""
+            self._entity_key_listeners[entity_key].remove(update_callback)
+            if not self._entity_key_listeners[entity_key]:
+                del self._entity_key_listeners[entity_key]
+
+        self._entity_key_listeners.setdefault(entity_key, []).append(update_callback)
+        return remove_listener
+
+    @callback
+    def async_update_listeners(
+        self, data: PassiveBluetoothDataUpdate[_T] | None
+    ) -> None:
+        """Update all registered listeners."""
+        # Dispatch to listeners without a filter key
+        for update_callback in self._listeners:
+            update_callback(data)
+
+        # Dispatch to listeners with a filter key
+        for key in self._entity_key_listeners:
+            if listeners := self._entity_key_listeners.get(key):
+                for update_callback in listeners:
+                    update_callback(data)
+
+    @callback
+    def _async_handle_bluetooth_event(
+        self,
+        service_info: BluetoothServiceInfo,
+        change: BluetoothChange,
+    ) -> None:
+        """Handle a Bluetooth event."""
+        self.name = service_info.name
+        self._last_callback_time = time.monotonic()
+        self._present = True
+
+        try:
+            new_data = self.update_method(service_info)
+        except Exception as err:  # pylint: disable=broad-except
+            self.last_update_success = False
+            self.logger.exception(
+                "Unexpected error updating %s data: %s", self.name, err
+            )
+            return
+
+        if not self.last_update_success:
+            self.last_update_success = True
+            self.logger.info("Processing %s data recovered", self.name)
+
+        if new_data:
```

Comment:
The return value of `update_method` isn't typed as optional.

Reply:
Removing the guard are we don't need it anymore 

f0158b6fb2f98e1a6cb1e96ecf316cb8d4bc2098